### PR TITLE
Roll out the test fetcher on push tests

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR rolls out the test_fetcher script on the merged PR in master, making it so that:
- circleCI only runs the tests that are affected by the diff between the master branch and the last commit
- non slow single and multi GPU tests only runs the tests that are affected by the diff between the master branch and the last commit

The last thing to add is a scheduled job to run all those tests daily to make sure we don't miss anything.

### CODE REVIEW

**Review by stas00 — APPROVED** [review]
Since this requires testing with actual changes and it's very hard to do in my head, I'm happy to have this rolled in and to report any issues if I run into at a later time, while being extra aware in the future PRs touching extended tests. 

Thank you for working on this super magic, @sgugger!

File: .circleci/config.yml [hunk 1]
```diff
@@ -579,6 +853,25 @@ workflows:
...
             - build_doc
             - deploy_doc: *workflow_filters
+    nightly:
+        triggers:
+            - schedule:
+                cron: "0 0 * * *"
+                filters:
+                    branches:
+                        only:
+                            - master
```
Comment (LysandreJik) [inline .circleci/config.yml:863]: Nice

File: .github/workflows/self-push.yml [hunk 2]
```diff
@@ -172,12 +208,24 @@ jobs:
...
+#      - name: Fetch the tests to run
+#        run: |
+#          python utils/tests_fetcher.py | tee test_preparation.txt
+#
+#      - name: Report fetched tests
+#        uses: actions/upload-artifact@v2
+#        with:
+#          name: test_fetched
+#          path: test_preparation.txt
+#
```
Comment (LysandreJik) [inline .github/workflows/self-push.yml:220]: Should all of these be cleaned up?

**Review by LysandreJik — APPROVED** [review]
This looks good to me! Thank you @sgugger!

File: .github/workflows/self-push.yml (hunk 2 above)
Comment (sgugger) [reply .github/workflows/self-push.yml:220]: No I put them here so that when we uncomment the TF tests, we have the same thing.

**Pull request merged**